src/neo4clj/query_builder.clj

```diff
@@ -62,81 +62,139 @@
+                 (conj rel-queries (cypher/relationship from-query to-query rel))
+                 (concat where-parts (remove nil? [from-where to-where])))))))
 
 (defn create-graph-query
   "Takes a graph representation and creates the nodes and relationship
   defined and returns any aliases specified in the representation"
   [{:keys [lookups nodes rels returns]}]
   (str/join " " (concat
                  (map #(lookup-node % false) lookups)
-                 (map #(create-node-query % false) nodes)
-                 (map create-graph-rel-query rels)
+                 [(str "CREATE " (str/join " CREATE " (first (generate-relation-queries
+                                                              (set (map :ref-id lookups))
+                                                              (reduce #(assoc %1 (:ref-id %2) %2) {} nodes)
+                                                              rels
+                                                              false))))]
                  (when-not (empty? returns)
-                   (vector (str "RETURN " (str/join "," (get-return-ref-ids returns))))))))
+                   (vector (str "RETURN " (str/join ", " (get-return-ref-ids returns))))))))
+
+(defn- lookup-unmatched-nodes
+  "Takes a list of node entries, known reference ids and a collection of relations.
+  Returns a collection of node queries and where quiries for the missing node reference ids."
+  [node-entries known-ref-ids rels]
+  (loop [unmatched-nodes []
+         remaining-rels rels
+         known-refs known-ref-ids]
+    (if (empty? remaining-rels)
+      (vector (remove nil? (map first unmatched-nodes))
+              (remove nil? (map second unmatched-nodes)))
+      (let [{:keys [from to]} (first rels)]
+        (recur (conj unmatched-nodes
+                     (node-reference known-refs node-entries from true true)
+                     (node-reference known-refs node-entries to true true))
+               (rest remaining-rels)
+               (conj known-refs (node-ref-id from) (node-ref-id to)))))))
+
+(defn- lookup-non-existing-graph-rel
+  "Takes a relation lookup representation and generates a bolt query with a
+  where part to test that the given relationship doesn't exists"
+  ([{:keys [from to] :as rel}]
+   (str "NOT " (cypher/relationship
+                (str "(" (or (:ref-id from) from) ")")
+                (str "(" (or (:ref-id to) to) ")")
+                (dissoc rel :ref-id)))))
+
+(defn- lookup-graph-single-matches-and-wheres
+  "Takes a list of node entries, known reference ids, relations to
+  check for non-existence and a collection of where queries.
+  Returns a query with missing node matches, non existent relation
+  wheres and other wheres given."
+  [node-entries known-ref-ids not-exists-rels where-parts]
+  (let [[unmatched-node-lookups unmatched-node-wheres] (lookup-unmatched-nodes node-entries known-ref-ids not-exists-rels)]
+    (str (when (not (empty? unmatched-node-lookups))
```

Comment:
Use `(not-empty ...` (without the questoinmark `?`) over `(not (empty? ...`.

This is true for several places in the code.

Reply:
Fixed by using not-empty and (when-not (empty?